web_embedding/cra-flutter/src/App/FlutterView/FlutterView.tsx

```diff
@@ -0,0 +1,79 @@
+import React, { useEffect, useRef, memo } from 'react'
+
+// The global _flutter namespace
+declare var _flutter: any
+
+// global promise is needed to avoid race conditions
+// when component is mounted and immediately unmounted,
+// e.g. in a React strict mode
+let engineInitializerPromise: Promise<any> | null = null
+
+interface FlutterViewProps {
+  assetBase?: string;
+  src?: string;
+  appLoaded?: (state: any) => void;
+}
+
+export const FlutterView: React.FC<FlutterViewProps> = memo(({
+  assetBase = '',
+  src = 'main.dart.js',
+  appLoaded,
+}) => {
+  const ref = useRef<HTMLDivElement>(null)
+
+  useEffect(() => {
+    let isRendered = true
+    const initFlutterApp = async () => {
+      if (!engineInitializerPromise) {
+        console.log('create Flutter engine initializer promise...')
+        engineInitializerPromise = new Promise<any>((resolve) => {
+          console.log('setup Flutter engine initializer...')
+          _flutter.loader.loadEntrypoint({
+            entrypointUrl: src,
+            onEntrypointLoaded: async (engineInitializer: any) => {
+              resolve(engineInitializer)
+            }
```

Comment:
Maybe you can simplify this as:

```suggestion
            onEntrypointLoaded: resolve
```

Reply:
(Also, why do you need to "leak" the engineInitializer out of this callback, you could potentially just have a `Promise<bool>` that resolves to 'true' once the whole async initialization procedure is done, right?)

Reply:
Component can be mounted and unmounted multiple times, but we can't call `loadEntrypoint` function multiple times as well -- it will result in an error. So we need a way to call `engineInitializer.initializeEngine` multiple times without re-creating `engineInitializer`. This means that the `engineInitializer` must be shared between different component re-renders

Reply:
Ohh, interesting, that might be a bug in the Angular version then, I don't remember attempting to preserve any of that :)

Reply:
It only happens if `loadEntrypoint` function is called a second time before the completion of the first call. So it is hard to detect. In this sample, special [StrictMode](https://react.dev/reference/react/StrictMode) component is used (exists by default for the newly created project), that is designed to find bugs in the components. For instance, it specifically re-renders components to find accidentally impure code.

Reply:
I have a related question here. If we call `initializeEngine` and `runApp` multiple times (like for the different re-renders), is it expected that `registerWith` is called multiple times for the web flutter plugins?

Reply:
@p-mazhnik I don't think we support very well multiple calls to `initializeEngine` and `runApp` right now, so the answer to your question is "undefined". However, according to the bootstrap of a Flutter Web app, `registerWith` gets called right before `runApp`:

* https://github.com/flutter/engine/blob/56b32827616b247ebf178ee33c08c3bb4022b141/lib/web_ui/lib/ui_web/src/ui_web/initialization.dart#L37-L43

Reply:
Turns out we can't have globally stored `engineInitializerPromise` :)
Just found that if we use previous `engineInitializer`, component _sometimes_ is not rendered in the element after remounting, even though code is loaded and interoperability works.

But looks like the issue with the multiple calls to `loadEntrypoint` is gone after I refactored code from
```js
_flutter.loader.loadEntrypoint({
   entrypointUrl: src,
   onEntrypointLoaded: async (engineInitializer: any) => {
    let appRunner = await engineInitializer.initializeEngine({
      hostElement: ref.current,
      assetBase: assetBase,
    });
    await appRunner.runApp();
  }
});
```
to
```js
const initFlutterApp = async () => {
  const engineInitializer = await new Promise<any>((resolve) => {
   _flutter.loader.loadEntrypoint({
      entrypointUrl: src,
      onEntrypointLoaded: resolve,
    })
  })

  const appRunner = await engineInitializer?.initializeEngine({
    hostElement: ref.current,
    assetBase: assetBase,
  })
  await appRunner?.runApp()
}
initFlutterApp()
```

As a small downside, our html page will have script tag injected on each render, but at least everything is working.

So I am going to remove global variable

Reply:
Should I create issues in the Flutter repo for my findings? I can probably provide some steps to reproduce.

However, I understand this is all new and not fully supported, so I am unsure about the actions that should be taken here and what is actually expected. All of it is related to the multiple calls of some function, which, as you said, are not well supported.
Anyway, I'll just leave them here, so I won't forget anything:

- If the standard Flutter initialization script is called two times 'too fast' (probably related to the `engineInitializer.initializeEngine` call and not to the `loadEntrypoint` as I thought before): 
`Uncaught TypeError: Cannot set properties of null (setting 'exports') at VM23285 canvaskit.js:219:18`
- If the standard Flutter initialization script is called two times 'too fast' (this one is probably related to the `appRunner.runApp` call and not to the `loadEntrypoint` as I thought before): 
`Uncaught TypeError: Cannot read properties of null (reading 'toString') at Object.HostNode_HostNode$create ... at Object.initializeEngineUi`
- if `engineInitializer` is reused, sometimes app is not rendered in the `hostElement`, but it looks like engine is still initialized (or the previous one is not cleaned up properly)
- nit: multiple calls to the `loadEntrypoint` inject multiple scripts on the page.
- (probably not related to Flutter itself, but still) `registerWith` called multiple times for the web plugins, causing different issues with the community plugins, because most of them currently assume that it will be called once.